File: OrderMatchingEngine/src/main/java/com/ndgndg91/ordermatchingengine/global/RequestAdviser.java
package com.ndgndg91.ordermatchingengine.global;

import lombok.extern.slf4j.Slf4j;
import org.springframework.context.support.DefaultMessageSourceResolvable;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.http.converter.HttpMessageNotReadableException;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;

import javax.validation.ConstraintViolation;
import javax.validation.ConstraintViolationException;
import java.time.LocalDateTime;
import java.util.stream.Collectors;

@Slf4j
@RestControllerAdvice
public class RequestAdviser {

    @ExceptionHandler(MethodArgumentNotValidException.class)
    public ResponseEntity<ApiError> validException(MethodArgumentNotValidException e) {
        final var message = e.getBindingResult().getFieldErrors().stream()
                .map(DefaultMessageSourceResolvable::getDefaultMessage)
                .collect(Collectors.joining("\t"));

        return ResponseEntity
                .status(HttpStatus.BAD_REQUEST)
                .body(new ApiError(HttpStatus.BAD_REQUEST.value(), message, null, LocalDateTime.now()));
    }

    @ExceptionHandler(HttpMessageNotReadableException.class)
    public ResponseEntity<ApiError> notReadableException(HttpMessageNotReadableException e) {
        return ResponseEntity
                .status(HttpStatus.BAD_REQUEST)
                .body(new ApiError(HttpStatus.BAD_REQUEST.value(), e.getMessage(), null, LocalDateTime.now()));
    }

    @ExceptionHandler(ConstraintViolationException.class)
    public ResponseEntity<ApiError> constraintViolationException(ConstraintViolationException e) {
        final var message = e.getConstraintViolations()
                .stream()
                .map(ConstraintViolation::getMessage)
                .findAny()
                .orElse("");

        log.info(message);
        return ResponseEntity
                .status(HttpStatus.BAD_REQUEST)
                .body(new ApiError(HttpStatus.BAD_REQUEST.value(), message, null, LocalDateTime.now()));
    }
}

File: OrderMatchingEngineKotlin/src/main/kotlin/com/ndgndg91/ordermatchingenginekotlin/global/RequestAdviser.kt
package com.ndgndg91.ordermatchingenginekotlin.global

import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.http.converter.HttpMessageNotReadableException
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice
import java.time.LocalDateTime
import javax.validation.ConstraintViolationException

@RestControllerAdvice
class RequestAdviser {
    val log: Logger = LoggerFactory.getLogger(RequestAdviser::class.java)

    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun notValidException(e: MethodArgumentNotValidException): ResponseEntity<ApiError> {
        val message = e.bindingResult.fieldErrors
            .asSequence()
            .map { it.defaultMessage }
            .joinToString("\t")

        return ResponseEntity
            .status(HttpStatus.BAD_REQUEST)
            .body(ApiError(HttpStatus.BAD_REQUEST.value(), message, null, LocalDateTime.now()))
    }

    @ExceptionHandler(HttpMessageNotReadableException::class)
    fun notReadableException(e: HttpMessageNotReadableException): ResponseEntity<ApiError> {
        return ResponseEntity
            .status(HttpStatus.BAD_REQUEST)
            .body(ApiError(HttpStatus.BAD_REQUEST.value(), e.message!!, null, LocalDateTime.now()))
    }

    @ExceptionHandler(ConstraintViolationException::class)
    fun constraintViolationException(e: ConstraintViolationException): ResponseEntity<ApiError> {
        val message = e.constraintViolations
            .asSequence()
            .map { it.messageTemplate }
            .joinToString("\t")

        log.info(message)
        return ResponseEntity
            .status(HttpStatus.BAD_REQUEST)
            .body(ApiError(HttpStatus.BAD_REQUEST.value(), message, null, LocalDateTime.now()))
    }



}